From b311ddc7c413d068b779275beb1e640f1ae6c1d8 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Wed, 8 Apr 2026 21:54:59 +0200
Subject: [PATCH] Fix potential undefined behavior in menu list obtaining

There is a case of modifying a list while foreaching over it, which can
potentially cause some issues.
---
 src/modulesmodel.cpp | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/src/modulesmodel.cpp b/src/modulesmodel.cpp
index 1d598bed..52db32f4 100644
--- a/src/modulesmodel.cpp
+++ b/src/modulesmodel.cpp
@@ -133,7 +133,7 @@ void ModulesModel::initMenuList(MenuItem *parent, const QList<KPluginMetaData> &
 
 QVariant ModulesModel::data(const QModelIndex &index, int role) const
 {
-    if (!index.isValid() || index.row() < 0 || index.row() >= rowCount()) {
+    if (!index.isValid()) {
         return {};
     }
 
@@ -276,14 +276,15 @@ QModelIndex ModulesModel::parent(const QModelIndex &index) const
 
 QList<MenuItem *> ModulesModel::childrenList(MenuItem *parent) const
 {
-    QList<MenuItem *> children = parent->children();
-    for (MenuItem *child : children) {
+    QList<MenuItem *> result;
+    for (MenuItem *child : std::as_const(parent->children())) {
         if (m_exceptions.contains(child)) {
-            children.removeOne(child);
-            children.append(child->children());
+            result.append(child->children());
+        } else {
+            result.append(child);
         }
     }
-    return children;
+    return result;
 }
 
 MenuItem *ModulesModel::parentItem(MenuItem *child) const
-- 
2.53.0

